### ISSUE TITLE
No module named "_mprsa"

### ISSUE THREAD

**Issue by matteopiccadaci** [git-n-pissed/mprsa#1]
Hi, it's my first time using an ESP32 so I don't know if I made any stupid mistake. I managed to build and flash the firmware (I used ESP IDF's docker container and the command you wrote in the ReadMe file). Micropython actually works and when I run the command help('modules') I get mprsa/\_\_init_\_ and mprsa/_asn1 among the others. When I run import mprsa, however, I get this error:
~~~ python
Traceback (most recent call last):
  File "<stdin>", line 25, in <module>
  File "mprsa/__init__.py", line 24, in <module>
ImportError: no module named '_mprsa'
~~~
I get this even when I run your test.py.

How can I fix it?
Thank you and best regards.

**Comment by git-n-pissed** [comment]
The `_mprsa` module (note the underscore) is defined in the C code, so if it is unavailable I would guess the `make` command used to build was incorrect.  Here is the command from [README.md](https://github.com/git-n-pissed/mprsa#readme) which I used as an example.  The `USER_C_MODULES` portion is the part I suspect is wrong for your error.

```bash
make -j8 BOARD=GENERIC_S3 USER_C_MODULES="$(pwd)/micropython/user_c_modules/mprsa/micropython.cmake" -C "$(pwd)/micropython/ports/esp32"
```

I've added a picture and some additional description to [README.md](https://github.com/git-n-pissed/mprsa#readme).  If you are unable to get it working, please post the exact commands you are using to build, the directory you are running the commands from, and the location of the `mprsa/user_c_modules` directory in relation to your MicroPython directory.

**Comment by matteopiccadaci** [comment]
I'll show you the complete procedure:

- I downloaded Micropython (v. 1.19.1) and your whole repository
- I put the folder USER_C_MODULES into the micropython directory (micropython-1.19.1/user_c_modules) and the mprsa folder (the one under port_modules) in micropython-1.19.1/ports/esp32/modules
![Screenshot 2023-01-12 alle 21 03 36](https://user-images.githubusercontent.com/95314113/212169303-462f9973-88ce-4ef6-b27f-4695f85545ac.png)
![Screenshot 2023-01-12 alle 21 04 31](https://user-images.githubusercontent.com/95314113/212169428-fc83b661-a9f5-4b43-aa7a-3431fed07f73.png)
- From the terminal, moved into the micropython-1.19.1 folder and I ran this docker command ```docker run --rm -v $PWD:/project -w /project -it espressif/idf:release-v4.4 ```
 (I used a docker container because I tried on MacOS (M1), Windows 10 and Linux (Zorin), but it gave me errors I honestly can't figure out) . As far as I know, this basically means that I'm using the folder micropython-1.19.1 as a volume for the container and, inside the container, the folder project will contain every file of micropython-1.19.1.

In fact, these are screenshots of the terminal window in which I ran the container:
![Screenshot 2023-01-12 alle 21 12 16](https://user-images.githubusercontent.com/95314113/212170912-90ec1b49-068d-4401-a5d3-c560e16ce122.png)

![Screenshot 2023-01-12 alle 21 14 46](https://user-images.githubusercontent.com/95314113/212171350-5a06d0f3-7883-47b6-9529-f694f48f6330.png)

- I then issued cd .. from project and ran ``` make -j8 BOARD=GENERIC_S3 USER_C_MODULES="$(pwd)/project/user_c_modules/mprsa/micropython.cmake" -C "$(pwd)/project/ports/esp32" ``` 
(very similar to your command, i just changed 'micropython' with 'project')

- Once finished, I exited the container, moved into micropython-1.19.1 and ran ``` idf.py -p /dev/tty.usbserial-0160DA0E flash ```  (as suggested into the container at the end of the make command). The flashing gets done correctly, but when I use Thonny IDE, I get the problem above.

**Comment by git-n-pissed** [comment]
> * I then issued cd .. from project and ran `make -j8 BOARD=GENERIC_S3 USER_C_MODULES="$(pwd)/project/user_c_modules/mprsa/micropython.cmake" -C "$(pwd)/project/ports/esp32"`
>   (very similar to your command, i just changed 'micropython' with 'project')
> * Once finished, I exited the container, moved into micropython-1.19.1 and ran `idf.py -p /dev/tty.usbserial-0160DA0E flash`  (as suggested into the container at the end of the make command). The flashing gets done correctly, but when I use Thonny IDE, I get the problem above.

Thanks for the detailed reply.  Everything looks mostly ok, but I suspect that changing the parent directory name from `micropython` to `project` might be causing some confusion with `make` or `cmake`.  Can you change the name of the parent directory back to `micropython` and try again?

Can you also confirm your hardware is an ESP32-S3?  The make command being used specifies the board is a `GENERIC_S3`.

One additional thing, you said you are using ESP-IDF v4.4.  This is a long shot, but the specific version I built with is `v4.4.3-239-gf90c12fc3c`.  I don't think this would be the issue, but it might be worth updating your ESP-IDF if you continue to have problems.  You can check your version by running `python esp-idf/tools/idf.py --version` from the directory that contains your ESP-IDF repo.

I build with WSL 1.0 on Windows 10 and am unable to test with Docker because my PC is too old and doesn't support the required virtualization.

**Comment by matteopiccadaci** [comment]
I tried to change the directory name and using ``` BOARD=GENERIC ``` instead of ``` BOARD=GENERIC_S3 ``` but I still get the error. ESP IDF in the container is version ``` v4.4.3-347-g9ee3c8337d ```. As for the version of the board, I don't really know because I borrowed it from my University: all I know is that when I flash it using esptool.py I read this ``` Chip is ESP32-D0WDQ6 (revision v1.0) ```, so my guess is that it is an old one.

**Comment by git-n-pissed** [comment]
The `D0WDQ6` v1.0 is the original chip used on the first ESP32; the ESP-WROOM-32, which, as you suspected, is an older board.  `BOARD=GENERIC` I do believe is correct, but the notes for that board  [here](https://github.com/micropython/micropython/blob/master/ports/esp32/boards/GENERIC/board.json#L19) and [here](https://github.com/micropython/micropython/blob/master/ports/esp32/boards/GENERIC/board.md) indicate it may need to be compiled with ESP-IDF 3.x :frowning_face:

FWIW, I changed my `micropython` directory name to `project`, rebuilt, and everything still worked fine for me.

Because you have come this far and you can actually flash your build, I would guess this is not the problem, but can you also confirm you have successfully built `mpy-cross` and `submodules` as discussed [here](https://docs.micropython.org/en/latest/develop/gettingstarted.html)?

**Comment by matteopiccadaci** [comment]
So I used the container release v. 3.3 and the build was successful but I still got the error. Tomorrow I'll try to flash using ESP-IDF v3.3 on Linux (on MacOS I flash using v 4.4.3, so maybe it could change). As for mpy-cross and submodules I confirm I can build them correctly (again only on container, because yesterday I tried building only mpy-cross and I got errors on both MacOS and Linux).

**Comment by git-n-pissed** [comment]
As an additional sanity check, you could try building the example user C module described [here](https://docs.micropython.org/en/latest/develop/cmodules.html).  It is in the `micropython/examples/usercmodule` directory.  If you can build successfully with and are able to import the module, that will tell us there is something different about mprsa.

Also, regarding ESP-IDF version, [this](https://github.com/micropython/micropython/discussions/10327#discussioncomment-4664097) was just posted today in the MicroPython Discussions.  So despite what the links I provided indicate, the MicroPython pre-built binaries are being compiled with ESP-IDF 4.2

**Comment by git-n-pissed** [comment]
@matteopiccadaci I found my old ESP-WROOM-32 today.  It has the same version of the same chip as the board you are using. I made a build for it with `mprsa` and I can confirm it works as expected.  I did discover one potential road block though, and that is that the default `gc.threshold()` is not sufficient for an ESP32.  I had to set call `gc.threshold(1024)` to get things the test to run correctly on the ESP32.  I'll update the documentation with this information.

I realize this doesn't help your build issue, but I can at least confirm that when built correctly and with an appropriate garbage collection threshold, the `mprsa` module will work on the ESP-WROOM-32.

For completeness, here is some information about how I built.  I'm using WSL 1.0 to build and I placed the `user_c_modules` and the `mprsa` directory as specified in "How to Build" portion of [README.md](https://github.com/git-n-pissed/mprsa/blob/main/README.md).  I ran the commands below from the `micropython` parent directory (the directory that contains the `micropython` directory).  The first command will be dependent on the location of your ESP-IDF:

```
source ../esp-idf/export.sh
make -j8 v=1 -C "$(pwd)/micropython/mpy-cross"
make -j8 "submodules" v=1 BOARD="GENERIC" -C "$(pwd)/micropython/ports/esp32"
make -j8 v=1 BOARD="GENERIC" USER_C_MODULES="$(pwd)/micropython/user_c_modules/micropython.cmake" -C "$(pwd)/micropython/ports/esp32"
```

And here is the commit and ESP-IDF version I was using:

MicroPython commit: [ac5934c96b66555ba8fe43465b34cb2b2a692ff0](https://github.com/micropython/micropython/commit/ac5934c96b66555ba8fe43465b34cb2b2a692ff0)
ESP-IDF version: `v4.4.3-239-gf90c12fc3c`

Finally, I flashed the device with esptool v4.2 using this command.  Note the `-z 0x1000` which is a flash offset required for this flavor of the ESP32:

```
python3 -m esptool --port /dev/ttyS3 --baud 921600 write_flash --erase-all -z 0x1000 micropython/ports/esp32/build-GENERIC/firmware.bin
```
